features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,1066 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------- Includes -----------------------------------------------------------
+
+#include "nvstore.h"
+#include "nvstore_int_flash_wrapper.h"
+#include "nvstore_shared_lock.h"
+#include "mbed_critical.h"
+#include "mbed_assert.h"
+#include "Thread.h"
+#include <string.h>
+#include <stdio.h>
+
+#if NVSTORE_ENABLED
+
+// --------------------------------------------------------- Definitions ----------------------------------------------------------
+
+#define PR_ERR printf
+#define PR_INFO printf
+#define PR_DEBUG printf
+
+#define MIN(a,b)            ((a) < (b) ? (a) : (b))
+#define MAX(a,b)            ((a) > (b) ? (a) : (b))
+
+#define DELETE_ITEM_FLAG  0x8000
+#define SET_ONCE_FLAG     0x4000
+#define HEADER_FLAG_MASK  0xF000
+
+#define MASTER_RECORD_KEY 0xFFE
+#define NO_KEY            0xFFF
+
+typedef struct {
+    uint16_t key_and_flags;
+    uint16_t length;
+    uint32_t mac;
+} record_header_t __attribute__((aligned(4)));
+
+
+#define OFFS_BY_KEY_AREA_MASK     0x80000000
+#define OFFS_BY_KEY_SET_ONCE_MASK 0x40000000
+#define OFFS_BY_KEY_FLAG_MASK     0xC0000000
+#define OFFS_BY_KEY_AREA_BIT_POS      31
+#define OFFS_BY_KEY_SET_ONCE_BIT_POS  30
+
+#define FLASH_MINIMAL_PROG_UNIT 8
+
+#define MASTER_RECORD_BLANK_FIELD_SIZE FLASH_MINIMAL_PROG_UNIT
+
+typedef struct {
+    uint16_t version;
+    uint16_t reserved1;
+    uint32_t reserved2;
+} master_record_data_t __attribute__((aligned(4)));
+
+#define MASTER_RECORD_SIZE sizeof(master_record_data_t)
+
+#define MEDITATE_TIME_MS 1
+
+typedef struct
+{
+    uint32_t address;
+    size_t   size;
+} nvstore_area_data_t;
+
+#if defined(NVSTORE_AREA_1_ADDRESS) || defined(NVSTORE_AREA_1_SIZE) ||\
+    defined(NVSTORE_AREA_2_ADDRESS) || defined(NVSTORE_AREA_2_SIZE)
+#define AREA_PARAMS_USER_CONFIG 1
+#endif
+
+#if AREA_PARAMS_USER_CONFIG
+#if !defined(NVSTORE_AREA_1_ADDRESS) || !defined(NVSTORE_AREA_1_SIZE) ||\
+    !defined(NVSTORE_AREA_2_ADDRESS) || !defined(NVSTORE_AREA_2_SIZE)
+#error Incomplete NVStore area configuration
+#endif
+#endif
+
+typedef enum {
+    AREA_STATE_NONE = 0,
+    AREA_STATE_EMPTY,
+    AREA_STATE_VALID,
+} area_state_e;
+
+#define INITIAL_CRC 0xFFFFFFFF
+
+// -------------------------------------------------- Local Functions Declaration ----------------------------------------------------
+
+// -------------------------------------------------- Functions Implementation ----------------------------------------------------
+
+// Safely increment an integer (depending on if we're thread safe or not)
+// Parameters :
+// value         - [IN]   Pointer to variable.
+// size          - [IN]   Increment.
+// Return        : Value after increment.
+int32_t safe_increment(uint32_t &value, uint32_t increment)
```

Comment:
Why not use the atomic increment directly? Seems an unnecessary helper.

Reply:
Right. This function had more logic in the past, but then evolved. Will change.